### ISSUE TITLE
Questions about the shape

### ISSUE THREAD

**Issue by zhongming12342** [M3RG-IITD/LaMO#1]
A great paper!!! I'm confused by the code such as for the Darcy and NS, where you mentioned that multi_heads is used in the NS. However, from the code, just one-head is used in this case. Another question is that your dimension about A and D are strange.

" N = self.d_state
        B, H, W, RD = x.shape
        K, R = self.A_logs.shape
        K, R, D = self.Ds.shape
        
        assert RD == R * D
        
        L = H * W
        KR = K * R
        _scan_mode = dict(cross2d=0, unidi=1, bidi=2, cascade2d=3)[self.scan_mode]

        initial_state = None
        if self.initial_state is not None:
            assert self.initial_state.shape[-1] == self.d_state
            initial_state = self.initial_state.detach().repeat(B, 1, 1, 1)
        xs = cross_scan_fn(x.view(B, H, W, RD), in_channel_first=False, out_channel_first=False, scans=_scan_mode, force_torch=self.scan_force_torch) # (B, H, W, 4, D)
        
        x_dbl = torch.einsum("b l k d, k c d -> b l k c", xs, self.x_proj_weight)
        if x_proj_bias is not None:
            x_dbl = x_dbl + x_proj_bias.view(1, -1, K, 1)
        
        dts, Bs, Cs = torch.split(x_dbl, [R, N, N], dim=3)
        xs = xs.contiguous().view(B, L, KR, D)
        dts = dts.contiguous().view(B, L, KR)
        Bs = Bs.contiguous().view(B, L, K, N)
        Cs = Cs.contiguous().view(B, L, K, N)
        
        if self.force_fp32:
            xs, dts, Bs, Cs = to_fp32(xs, dts, Bs, Cs)

        As = -self.A_logs.to(torch.float).exp().view(KR)
        Ds = self.Ds.to(torch.float).view(KR, D)
        dt_bias = self.dt_projs_bias.view(KR)

        if self.force_fp32:
            xs, dts, Bs, Cs = to_fp32(xs, dts, Bs, Cs)
        
        ys, final_state = mamba_chunk_scan_combined(
            xs, dts, As, Bs, Cs, chunk_size = self.chunk_size, D=Ds, dt_bias=dt_bias, 
            initial_states=initial_state, dt_softplus=True, return_final_states=True,
        )
                
        y: torch.Tensor = cross_merge_fn(ys.view(B, H, W, K, RD), in_channel_first=False, out_channel_first=False, scans=_scan_mode, force_torch=self.scan_force_torch)"

A should be the dimension of [k_group,d_inner,d_state]? and D is [k_group, d_state]? and why you reshape "xs = xs.contiguous().view(B, L, KR, D)"? what does this mean?

**Comment by Karn3003** [comment]
Hi,

Thanks for the kind words.

**“I'm confused by the code such as for the Darcy and NS, where you mentioned that multi_heads is used in the NS.”**

As for the regular grid, we are doing **4-scan SSM**, so the only “heads’’ here are the **K groups** produced by the 2D cross scans (`cross2d`): **up, down, left, right → K = 4**.  
It’s *one head per scan direction*.  
These are **scan-heads** in a regular grid.

A few notations for the next question are as follows:

- **N = self.d_state** — the SSM state size.  
- **dt_rank = R** — low-rank factorization rank used to produce **R parallel SSM streams** per scan-group.  
- **D = d_inner // R** — the per-rank channel slice.  
- **K** — number of scan groups (directions). With `scan_mode="cross2d"`, **K = 4**.
---

**“Another question is that your dimension about A and D are strange.”**

They are not `(K, d_inner, d_state)` or `(K, d_state)`.  
In this implementation, each SSM **stream** is parameterized by:

- a **scalar A** per (scan-group, rank),  
- a **D-dimensional vector** per (scan-group, rank).

So the actual shapes are:

- `A_logs: (K, R)` → after `exp` and sign flip → `As: (K*R,)`  
- `Ds: (K, R, D)` → reshaped to `Ds: (K*R, D)`

This matches the fused SSM kernel which expects **KR independent SSM streams**, each with:

- a scalar `A`  
- a diagonal vector `D`  
- a per-stream dt (from `dts`)  
- shared `B, C` within each K (broadcasted over R)
---
**“and why you reshape `xs = xs.contiguous().view(B, L, KR, D)`? what does this mean?”**

After the cross-scan, `xs` has shape:

(B, H, W, K, R*D)

Flattening spatial dims (`H*W = L`) and splitting the channel `R*D` into `(R, D)` gives:

xs → (B, L, K, R, D)

Then we merge `(K, R)` into a single dimension:

(B, L, K, R, D) → (B, L, K*R, D)

The fused kernel `mamba_chunk_scan_combined` runs **KR parallel SSM streams**, one per `(scan-direction, rank)` pair.  
Reshaping into `(B, L, KR, D)` gives the exact format the kernel expects.

After the scan finishes, we reshape back and undo the packing.

---

For a better understanding of Mamba 4D, please refer to the following repository, as our implementation is adapted from it.

https://github.com/MzeroMiko/VMamba
